pkg/metrics/metrics.go

```diff
@@ -7,17 +7,144 @@ import (
+		Name: "vertex_pending_messages",
+		Help: "Pending messages of the vertex",
+	}, []string{"vertex", "pipeline"})
+)
+
+// timestampedPending is a helper struct to wrap a pending number and timestamp pair
+type timestampedPending struct {
+	pending int64
+	// timestamp in seconds
+	timestamp int64
+}
+
+type metricsServer struct {
+	vertex    *dfv1.Vertex
+	rater     isb.Ratable
+	lagReader isb.LagReader
+	// pendingLookbackSeconds is the look back seconds for pending calculation
+	pendingLookbackSeconds int64
+	refreshInterval        time.Duration
+	// pendingInfo stores a list of pending/timestamp(seconds) information
+	pendingInfo *sharedqueue.OverflowQueue[timestampedPending]
+}
+
+type Option func(*metricsServer)
+
+func WithRater(r isb.Ratable) Option {
+	return func(m *metricsServer) {
+		m.rater = r
+	}
+}
+
+func WithLagReader(r isb.LagReader) Option {
+	return func(m *metricsServer) {
+		m.lagReader = r
+	}
+}
+
+func WithRefreshInterval(d time.Duration) Option {
+	return func(m *metricsServer) {
+		m.refreshInterval = d
+	}
+}
+
+func WithPendingLookbackSeconds(seconds int64) Option {
+	return func(m *metricsServer) {
+		m.pendingLookbackSeconds = seconds
+	}
+}
+
+func NewMetricsServer(vertex *dfv1.Vertex, opts ...Option) *metricsServer {
+	m := new(metricsServer)
+	m.vertex = vertex
+	m.refreshInterval = 5 * time.Second // Default refersh interval
+	m.pendingLookbackSeconds = 180      // Default
+	for _, opt := range opts {
+		if opt != nil {
+			opt(m)
+		}
+	}
+	if m.lagReader != nil {
+		m.pendingInfo = sharedqueue.New[timestampedPending](1800)
+	}
+	return m
+}
+
+func (ms *metricsServer) rateAndPending(ctx context.Context) {
+	if ms.lagReader == nil && ms.rater == nil {
+		return
+	}
+	labels := map[string]string{"vertex": ms.vertex.Spec.Name, "pipeline": ms.vertex.Spec.PipelineName}
+	log := logging.FromContext(ctx)
+	ticker := time.NewTicker(ms.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if ms.rater != nil {
+				if r, err := ms.rater.Rate(ctx); err != nil {
+					log.Errorw("failed to get processing rate", zap.Error(err))
+				} else {
+					if r != isb.RateNotAvailable {
+						processingRate.With(labels).Set(r)
+					}
+				}
+			}
+			if ms.lagReader != nil {
+				if pending, err := ms.lagReader.Pending(ctx); err != nil {
+					log.Errorw("failed to get pending messages", zap.Error(err))
+				} else {
+					if pending != isb.PendingNotAvailable {
+						now := time.Now().Unix()
+						ts := timestampedPending{pending: pending, timestamp: now}
+						ms.pendingInfo.Append(ts)
+						// Calculate avg pending
+						items := ms.pendingInfo.Items()
+						total := pending
+						num := int64(1)
+						for i := len(items) - 2; i >= 0; i-- {
+							if now-items[i].timestamp < ms.pendingLookbackSeconds {
+								total += items[i].pending
+								num++
+							} else {
+								break
+							}
+						}
+						pendingMessages.With(labels).Set(float64(total / num))
+					}
```

Comment:
should we emit these metrics to prometheus too?

Reply:
I assume you meant shall we display these metrics (they are already exposed/emmited) - This is the tricky part.

The `pending_messages` and `processing_rate` represent the vertex level metrics, not a pod, but we are exposing them from every individual pod, this will be a problem when displaying in the dashboard, that we should only display the metrics from one pod (or avg).

Reply:
This the reason of the discussion about should we only expose these metrics from 1 pod.

Reply:
we could expose from all pods, the numbers should be more or less the same.. we can let the user configure how they want to display avg/max/min etc?

Reply:
I agree with Vigith that we should expose from all the pods and let the user decide when they are displaying these metrics on the type of stat to be used. We just need to call this out specifically in our doc that these numbers are vertex level and should roughly be the same and the aggregation we need to use is just `avg/max/min`?